packages/react-router/lib/hooks.tsx

```diff
@@ -972,29 +971,14 @@ export function useBlocker(shouldBlock: boolean | BlockerFunction): Blocker {
     [basename, shouldBlock]
   );
 
-  // This effect is in charge of blocker key assignment and deletion (which is
-  // tightly coupled to the key)
+  // register and unregister router blocker using blocker unique key
   React.useEffect(() => {
-    let key = String(++blockerId);
-    setBlockerKey(key);
-    return () => router.deleteBlocker(key);
-  }, [router]);
-
-  // This effect handles assigning the blockerFunction.  This is to handle
-  // unstable blocker function identities, and happens only after the prior
-  // effect so we don't get an orphaned blockerFunction in the router with a
-  // key of "".  Until then we just have the IDLE_BLOCKER.
-  React.useEffect(() => {
-    if (blockerKey !== "") {
-      setBlocker(router.getBlocker(blockerKey, blockerFunction));
-    }
+    router.getBlocker(blockerKey, blockerFunction);
+    return () => router.deleteBlocker(blockerKey);
```

Comment:
I think this is problematic since if the `blockerFunction` changes identities we will delete and recreate the blocker, which would lose any currently `blocked` state.  You can see this in `examples/navigation-blocking` which creates a callback based on the controlled input value such as

```js
let shouldBlock = React.useCallback(() => value !== "", [value]);
let blocker = useBlocker(shouldBlock);
```

That means that if you get into a blocked state (enter input value `12345` and click a link) and then you delete a few chars (leaving it at `123`) you lose your blocked state entirely.  

The current approach decouples blocker creation/deletion from the updating of the `blockerFunction` should it change, which alleviates this issue.

Reply:
I did end up removing `blocker`/`setBlocker` in https://github.com/remix-run/react-router/pull/10656 and it seems to work well

Reply:
> That means that if you get into a blocked state (enter input value `12345` and click a link) and then you delete a few chars (leaving it at `123`) you lose your blocked state entirely.

makes sense. If the problem is only the unmount bit we can probably can simply split the use effect in 2 with the unmount with dependency array containing only the key right?

Reply:
Potentially.  Let's let `6.14.1` get released and merged to `dev` which will cause some conflicts in here, then we can test out the `useMemo` addition once those are resolved